### PR TITLE
`functools`: Add typing for decorated function arguments in cache API

### PR DESCRIPTION
**Pull request by asvetlov**
`functools.cache()` and `functools.lru_cache()` are affected.

### CODE REVIEW

**Comment by asvetlov** [comment]
Instancemethod seems working, but classmethod does not.

**Comment by AlexWaygood** [comment]
There was a previous attempt at doing something like this in #7771 that was quite promising, but the author closed the PR

**Comment by AlexWaygood** [comment]
See this issue for some discussion on this, as well:

- #6347

File: stdlib/functools.pyi [hunk 1]
```diff
@@ -45,22 +46,32 @@ class _CacheInfo(NamedTuple):
...
     def cache_clear(self) -> None: ...
-    def __copy__(self) -> _lru_cache_wrapper[_T]: ...
-    def __deepcopy__(self, __memo: Any) -> _lru_cache_wrapper[_T]: ...
+    def __copy__(self) -> _lru_cache_wrapper[_P, _T]: ...
+    def __deepcopy__(self, __memo: Any) -> _lru_cache_wrapper[_P, _T]: ...
+    if sys.version_info >= (3, 8):
+        @overload
+        def __get__(self, __instance: None, __owner: type[_S] | None = ...) -> _lru_cache_wrapper[_P, _T]: ...
+        @overload
+        def __get__(self, __instance: _S, __owner: type[_S] | None = ...) -> Callable[Concatenate[_S, _P], _T]: ...
```
Comment (WMOkiishi) [inline stdlib/functools.pyi:60]: It seems you have this backward. I think it should be this:
```suggestion
        def __get__(self: _lru_cache_wrapper[Concatenate[_S, _Q], _T], __instance: _S, __owner: type[_S] | None = ...) -> Callable[_Q, _T]: ...
```
Where `_Q = ParamSpec("_Q")`. You'd also want to change the one for 3.8 below.

**Comment by WMOkiishi** [comment]
@asvetlov any chance you could see what effect my suggested changes would have on the primer output? I can, of course, make my own PR if you don't have the bandwidth at the moment, but I figured it would be polite to bring it up here first.

**Comment by jakkdl** [comment]
> @asvetlov any chance you could see what effect my suggested changes would have on the primer output? I can, of course, make my own PR if you don't have the bandwidth at the moment, but I figured it would be polite to bring it up here first.

asvetlov seems to have dropped this, so you can go ahead. I might try to give it a stab myself otherwise.

**Comment by WMOkiishi** [comment]
> I might try to give it a stab myself otherwise.

Go for it! I don't think I really have the time to work on that right now.

**Comment by srittau** [comment]
There are now non-trivial merge conflicts.

**Comment by jakkdl** [comment]
I think the PR can be closed.

**Pull request closed without merging**